### PR TITLE
Remote queries: Handle nested queries

### PR DESCRIPTION
**Pull request by aeisenberg**
This change allows remote queries to run a query from a directory that
is not in the root of the qlpack.

The change is the following:

1. walk up the directory hierarchy to check for a non-local qlpack.yml
2. Copy over the files as before, but keep track of the relative
   location of the query compared to the location of the qlpack.yml.
3. Change the defaultSuite of the qlpack.yml so that _only_ this query
   is run as part of the default query.

Also, this adds a new integration test to ensure the nested query is
packaged appropriately.

@rneatherway note that in order for this change to work on the server, the default query suite of the pack needs to be invoked, not the query itself.

## Checklist

- [x] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [x] Issues have been created for any UI or other user-facing changes made by this pull request.
- [x] `@github/docs-content-codeql` has been cc'd in all issues for UI or other user-facing changes made by this pull request.

### CODE REVIEW

File: extensions/ql-vscode/src/vscode-tests/cli-integration/run-remote-query.test.ts [hunk 1]
```diff
@@ -180,15 +236,37 @@ describe('Remote queries', function() {
...
     return Uri.file(path.join(baseDir, file));
   }
 
   function printDirectoryContents(dir: string) {
-    console.log(`DIR ${dir}`);
-    if (!fs.existsSync(dir)) {
-      console.log(`DIR ${dir} does not exist`);
-    }
-    fs.readdirSync(dir).sort().forEach(f => console.log(`  ${f}`));
+    dir;
```
Comment (shati-patel) [inline extensions/ql-vscode/src/vscode-tests/cli-integration/run-remote-query.test.ts:264]: Do we need to worry about this `Expression has no effect` code scanning warning? 🔍

File: extensions/ql-vscode/src/vscode-tests/cli-integration/run-remote-query.test.ts [hunk 2]
```diff
@@ -180,15 +236,37 @@ describe('Remote queries', function() {
...
+    expect(qlPack).to.deep.equal({
+      name: packName,
+      version: packVersion,
+      dependencies: {
+        'codeql/javascript-all': '*',
+      },
+      library: false,
+      defaultSuite: [{
+        description: 'Query suite for remote query',
+        query: queryPath,
```
Comment (shati-patel) [inline extensions/ql-vscode/src/vscode-tests/cli-integration/run-remote-query.test.ts:254]: The Windows tests are unhappy about path separators: https://github.com/github/vscode-codeql/runs/4259501142?check_suite_focus=true#step:9:224

Not sure if the problem comes from `queryPath` here, or from the logic in `run-remote-query.ts` itself....

File: extensions/ql-vscode/src/run-remote-query.ts [hunk 3]
```diff
@@ -413,3 +429,23 @@ export async function attemptRerun(
...
+/**
+ * Updates the default suite of the query pack. This is used to ensure
+ * only the specified query is run.
+ *
+ * @param queryPackDir The directory containing the query pack
+ * @param packRelativePath The relative path to the query pack from the root of the query pack
+ */
+async function fixDefaultSuite(queryPackDir: string, packRelativePath: string): Promise<void> {
+  const packPath = path.join(queryPackDir, 'qlpack.yml');
+  const qlpack = (await yaml.safeLoad(await fs.readFile(packPath, 'utf8'))) as QlPack;
```
Comment (shati-patel) [inline extensions/ql-vscode/src/run-remote-query.ts:442]: (Minor question: does the first `await` on this line do anything? I don't think we await `yaml.safeLoad` anywhere else, so just checking I understand 😅 )

**Review by shati-patel — COMMENTED** [review]
Thanks for doing this! ✨ Looks good to me, though it would be helpful to test it once the server-side changes are in place too 😎

File: extensions/ql-vscode/CHANGELOG.md [hunk 4]
```diff
@@ -10,6 +10,7 @@
 - Add a _CodeQL: Preview Query Help_ command to generate Markdown previews of `.qhelp` query help files. This command should only be run in trusted workspaces. See https://codeql.github.com/docs/codeql-cli/testing-query-help-files for more information about query help. [#988](https://github.com/github/vscode-codeql/pull/988)
 - Make "Open Referenced File" command accessible from the active editor menu. [#989](https://github.com/github/vscode-codeql/pull/989)
 - Allow query result locations with 0 as the end column value. These are treated as the first column in the line. [#1002](https://github.com/github/vscode-codeql/pull/1002)
+- Change remote query support so that query files that are not in the root of the query pack can be uploaded and run. [#1009](https://github.com/github/vscode-codeql/pull/1009)
```
[the remark this answers is not in the document]
Comment (aeisenberg) [reply extensions/ql-vscode/CHANGELOG.md:13]: Sure thing.

File: extensions/ql-vscode/src/vscode-tests/cli-integration/run-remote-query.test.ts (hunk 1 above)
Comment (aeisenberg) [reply extensions/ql-vscode/src/vscode-tests/cli-integration/run-remote-query.test.ts:264]: oops...shouldn't have committed that.

File: extensions/ql-vscode/src/vscode-tests/cli-integration/run-remote-query.test.ts (hunk 2 above)
Comment (aeisenberg) [reply extensions/ql-vscode/src/vscode-tests/cli-integration/run-remote-query.test.ts:254]: hmmmm...let me investigate.

File: extensions/ql-vscode/src/run-remote-query.ts (hunk 3 above)
Comment (aeisenberg) [reply extensions/ql-vscode/src/run-remote-query.ts:442]: No...I don't think I need it. I'll remove.

File: extensions/ql-vscode/src/vscode-tests/cli-integration/run-remote-query.test.ts (hunk 2 above)
Comment (aeisenberg) [reply extensions/ql-vscode/src/vscode-tests/cli-integration/run-remote-query.test.ts:254]: Made a fix. Glad we run these tests on multiple platforms.

**Comment by aeisenberg** [comment]
Really quite confused why windows is still failing. I'll have to fire up my vm to try this out.

**Comment by aeisenberg** [comment]
The failing tests (at least some of them) are due to a bug in the CLI. The `defaultSuite` field in `qlpack.yml` is not serialized correctly on windows. The fix will make it into the next release of the CLI (likely v2.7.3) and I will augment the tests so that they the failing ones will not run on windows cli <= 2.7.2.

File: extensions/ql-vscode/src/run-remote-query.ts [hunk 5]
```diff
@@ -413,3 +434,23 @@ export async function attemptRerun(
     void showAndLogErrorMessage(error);
   }
 }
+
+/**
+ * Updates the default suite of the query pack. This is used to ensure
+ * only the specified query is run.
+ *
+ * @param queryPackDir The directory containing the query pack
+ * @param packRelativePath The relative path to the query pack from the root of the query pack
```
Comment (shati-patel) [inline extensions/ql-vscode/src/run-remote-query.ts:443]: The wording looks a bit confusing here 😅  Should this be... ?
```suggestion
 * @param packRelativePath The relative path to the query from the root of the query pack
```

File: extensions/ql-vscode/src/vscode-tests/cli-integration/run-remote-query.test.ts [hunk 6]
```diff
@@ -180,6 +238,41 @@ describe('Remote queries', function() {
     }
   });
 
+  function verifyQlPack(qlpackPath: string, queryPath: string, packName: string, packVersion: string, pathSerializationBroken: boolean) {
+    const qlPack = yaml.safeLoad(fs.readFileSync(qlpackPath, 'utf8')) as QlPack;
+
+    if (pathSerializationBroken) {
+      // the path serialization is broken, so we force it to be the path in the pack to be same as the query path
+      qlPack.defaultSuite![1].query = queryPath;
+    }
```
Comment (shati-patel) [inline extensions/ql-vscode/src/vscode-tests/cli-integration/run-remote-query.test.ts:247]:
```suggestion
    if (pathSerializationBroken) {
      // the path serialization is broken, so we force the path in the pack to be same as the query path
      qlPack.defaultSuite![1].query = queryPath;
    }
```

or something like this?

**Review by shati-patel — APPROVED** [review]
Thanks for figuring out the problem with the Windows tests! 🎉 

Looks good from my point of view now, though we'll need to hold off until the server-side changes are in place too.

File: extensions/ql-vscode/src/vscode-tests/cli-integration/run-remote-query.test.ts (hunk 6 above)
Comment (aeisenberg) [reply extensions/ql-vscode/src/vscode-tests/cli-integration/run-remote-query.test.ts]: Sure...I think that was laziness from me on relying on co-pilot too much.:)

**Comment by rneatherway** [comment]
We were just testing this and right now it fails because the name of the pack isn't stable. Do you think that we can just do this?

```diff
diff --git a/extensions/ql-vscode/src/run-remote-query.ts b/extensions/ql-vscode/src/run-remote-query.ts
index 0689ea5..212c4a8 100644
--- a/extensions/ql-vscode/src/run-remote-query.ts
+++ b/extensions/ql-vscode/src/run-remote-query.ts
@@ -180,14 +180,8 @@ async function generateQueryPack(cliServer: cli.CodeQLCliServer, queryFile: stri
  */
 async function ensureQueryPackName(queryPackDir: string) {
   const pack = yaml.safeLoad(await fs.readFile(path.join(queryPackDir, 'qlpack.yml'), 'utf8')) as { name: string; };
-  if (!pack.name || !pack.name.includes('/')) {
-    if (!pack.name) {
-      pack.name = 'codeql-remote/query';
-    } else if (!pack.name.includes('/')) {
-      pack.name = `codeql-remote/${pack.name}`;
-    }
-    await fs.writeFile(path.join(queryPackDir, 'qlpack.yml'), yaml.safeDump(pack));
-  }
+  pack.name = 'codeql-remote/query';
+  await fs.writeFile(path.join(queryPackDir, 'qlpack.yml'), yaml.safeDump(pack));
 }
```

**Comment by aeisenberg** [comment]
Oh...apologies. I made a fix for the name yesterday, but neglected to push it up. It's slightly different than what you suggest, but it's the same basic idea.

**Comment by rneatherway** [comment]
> Oh...apologies. I made a fix for the name yesterday, but neglected to push it up. It's slightly different than what you suggest, but it's the same basic idea.

No problem! I've tested it and that version works too :-)

**Review by rneatherway — APPROVED** [review]
LGTM, just need to update that last test to match the pack name

**Review by rneatherway — APPROVED** [review]
:+1:

**Pull request merged**